### ISSUE TITLE
Gitcoin: 7) Port an Existing Ethereum dApp to Polyjuice

### ISSUE THREAD

**Issue by matt-quinn** [nervosnetwork/grants#8]
This task is capped at maximum 125 submissions; so be in the first 125 valid submissions to get your payout!

# 7. Port an Existing Ethereum dApp to Polyjuice

In this tutorial you will learn how to port an existing browser Ethereum application to run on Nervos' EVM compatible [Layer 2](https://github.com/Kuzirashi/gw-gitcoin-instruction/tree/master/src/conceptual-explainers/structure.md#layer-1--layer-2).

[MetaMask](https://github.com/Kuzirashi/gw-gitcoin-instruction/tree/master/src/conceptual-explainers/wallets.md#metamask) is the primary method of interacting with dApps on Nervos. The over user experience will be very similar to existing Ethereum. The one big difference is that the MetaMask confirmation dialog will ask them to sign a hex string instead of displaying transaction. This is normal behavior when using MetaMask on Nervos since MetaMask is currently not capable of deciphering a [Godwoken](https://github.com/Kuzirashi/gw-gitcoin-instruction/tree/master/src/conceptual-explainers/frameworks.md#godwoken) transaction.

The eventual goal of [Polyjuice](https://github.com/Kuzirashi/gw-gitcoin-instruction/tree/master/src/conceptual-explainers/frameworks.md#polyjuice) is 100% compatibility with existing EVM smart contracts. However, this goal is still being worked on, and there will always be a few differences in the setup of the development environment and the tooling and frameworks that are used.

## Task Instructions

> Note: Before starting the tasks, it is recommended that you review the [Task Submission](#task-submission) section so you know what materials you will need to provide to judges to review your task submission.

In this task, you will need to port over an existing Ethereum dApp to Nervos' Layer 2. Our examples will use the Simple Storage smart contract, but your submission must be significantly different. Using the Simple Storage smart contract is allowed, but only the end result of your dApp is significant different than the example. If it your application is too similar to the example, it will be rejected by the judges. Our general suggestion is that you take an existing Ethereum dApp that you are already familiar with, then make the changes required to port it to Nervos.

In this guide we assume you have application that is using **Web3.js**. Instructions for Ethers.js can be found in ["Additional Tutorials and Code Resources"](#additional-tutorials-and-code-resources).

> Important: The example in this task uses the Simple Storage smart contract, but you cannot clone this code for your task. **In order to get credit for this task you must use a different project.**

### Prerequisites

Your MetaMask wallet should be installed and configured with the same Ethereum accounts from the previous tasks. Your MetaMask Ethereum account should still be funded on Layer 2 from the previous tasks.

### 1. Setup the Godwoken Testnet Network in MetaMask

Your MetaMask wallet will need to be configured to communicate with the Godwoken Layer 2 network. To do this, you will need to configure a new custom RPC. From the network selection dropdown, select "Custom RPC".

<img src="https://raw.githubusercontent.com/Kuzirashi/gw-gitcoin-instruction/master/src/images/metamask-network-menu.png" alt="MetaMask Network Menu" style="height: 300px; border: 1px solid #eeeeee" />

From there you will be presented with a form to specify the network settings.

<img src="https://raw.githubusercontent.com/Kuzirashi/gw-gitcoin-instruction/master/src/images/metamask-networks.png" alt="MetaMask Add Network" style="height: 300px; border: 1px solid #eeeeee" />

Enter the following details.

```txt
Network Name: Godwoken Testnet
RPC URL: https://godwoken-testnet-web3-rpc.ckbapp.dev
Chain ID: 71393
Currency Symbol: <Leave Empty>
Block Explorer URL: <Leave Empty>
```

After MetaMask is configured you may see a zero balance even after you have deposited funds into this Ethereum address on Layer 2. Don't be alarmed by this. Later on we will show you how to setup your application to call `web3.eth.getBalance` with your Ethereum address to query for your balance.

### 2. View the Ethereum Demo Application

A simple Ethereum demo application has been built to use the Simple Storage to read and write number values. We will walk through this Ethereum application, and demonstrate the changes needed to make this run on Nervos' Layer 2.

We recommend that you clone the source code and follow along with our explanations so you become familiar with the process invovled in porting an application.

> Note: Remember, the application you submit must be different from our example. This guide will walk you through the process, but you must port over a different application to submit your task.

First, we need to clone the repo. This contains the Ethereum version of the application before it has been updated to support Godwoken.

Create `~/projects` directory if it doesn't exist.

Linux/MacOS:

```sh
mkdir -p ~/projects
```

Windows (PowerShell):

```ps
New-Item -ItemType Directory -Force -Path ~/projects
```

Enter the `~/projects` directory and clone the repository.

```sh
cd ~/projects
git clone https://github.com/Kuzirashi/blockchain-workshop.git -b ethereum-simple blockchain-workshop-ethereum-simple
cd blockchain-workshop-ethereum-simple
```

Next, install the dependencies, build the smart contracts, and start Ganache to run a local Ethereum development chain.

```sh
cd ~/projects/blockchain-workshop-ethereum-simple
yarn
yarn build
yarn start:ganache
```

Ganache should now be running and creating blocks.

Switch back to your web browser. Open MetaMask and switch your network to `Localhost 8545`. It should switch and connect without giving an error now that Ganache is running.

<img src="https://raw.githubusercontent.com/Kuzirashi/gw-gitcoin-instruction/master/src/images/metamask-localhost.png" alt="MetaMask Localhost 8545" style="height: 300px; border: 1px solid #eeeeee"/>

In a second terminal, start the UI server.

```sh
cd ~/projects/blockchain-workshop-ethereum-simple
yarn ui
```

The server should now be started, and you can open a browser tab to http://localhost:3000 to view the dApp UI!

<img src="https://raw.githubusercontent.com/Kuzirashi/gw-gitcoin-instruction/master/src/images/simple-storage-application.png" alt="Simple Storage Application" style="height: 300px; border: 1px solid #eeeeee" />

### 3. Install Polyjuice Dependencies

Now we will begin porting this Ethereum application to use Nervos' Layer 2. The first step is to install the required dependencies for working with Godwoken and Polyjuice. Use the following command in the main project directory to install these dependencies in your application.

```js
cd ~/projects/blockchain-workshop-ethereum-simple
yarn add @polyjuice-provider/web3@0.0.1-rc7 nervos-godwoken-integration@0.0.6
```

- `@polyjuice-provider/web3` is a custom Polyjuice web3 provider. It is required for interaction with Nervos' Layer 2 smart contracts.
- `nervos-godwoken-integration` is a tool that can generate Polyjuice address based on your Ethereum address. You might be required to use Polyjuice address if you store values mapped to addresses in your contracts.

### 4. Configure the Web3 Provider for the Polyjuice Web3 Provider

The next step is to configure the Polyjuice Web3 Provider for the application. This replaces the normal web3 provider that may be currently in use for Ethereum with one for the Godwoken Testnet.

The example below is just for demonstration of how your configuration file could look. You will not find it in any of the example project source files.

```js
export const CONFIG = {
    WEB3_PROVIDER_URL: 'http://localhost:8545'
};
```

is changed to:

```js
export const CONFIG = {
    WEB3_PROVIDER_URL: 'https://godwoken-testnet-web3-rpc.ckbapp.dev'
};
```

> Note: Your configuration might be using `localhost:8545`, or some other settings for some other service like Infura.

The two other values needed for the Polyjuice Web3 Provider are the `ROLLUP_TYPE_HASH` and the `ETH_ACCOUNT_LOCK_CODE_HASH`. Both of these values are constants needed for the Godwoken Testnet. A full list of the various values for the Godwoken Testnet can be found [here](https://github.com/jjyr/godwoken-testnet).

We will add these values to our configuration object below. These values will be needed by our project, so we are creating a new config file called `config.ts` to our source code.

Here is the contents of the file `~/projects/blockchain-workshop-ethereum-simple/src/config.ts` that we just created:

```js
export const CONFIG = {
    WEB3_PROVIDER_URL: 'https://godwoken-testnet-web3-rpc.ckbapp.dev',
    ROLLUP_TYPE_HASH: '0x4cc2e6526204ae6a2e8fcf12f7ad472f41a1606d5b9624beebd215d780809f6a',
    ETH_ACCOUNT_LOCK_CODE_HASH: '0xdeec13a7b8e100579541384ccaf4b5223733e4a5483c3aec95ddc4c1d5ea5b22'
};
```

We will use these config values in a moment, but first we need to import a few dependencies. We will update the main UI in the file `~/projects/blockchain-workshop-ethereum-simple/src/ui/app.tsx`.

Next, we add the following lines in the main dependency importation section of the file.

```js
import { PolyjuiceHttpProvider } from '@polyjuice-provider/web3';
import { CONFIG } from '../config';
```

This imports the Polyjuice Web3 Provider, which we will use in a moment, and the config file that we just created.

Next we prepare a few constants, create the Polyjuice Provider, and use the Polyjuice Provider with a Web3 instance.

```js
const godwokenRpcUrl = CONFIG.WEB3_PROVIDER_URL;
const providerConfig = {
    rollupTypeHash: CONFIG.ROLLUP_TYPE_HASH,
    ethAccountLockCodeHash: CONFIG.ETH_ACCOUNT_LOCK_CODE_HASH,
    web3Url: godwokenRpcUrl
};
const provider = new PolyjuiceHttpProvider(godwokenRpcUrl, providerConfig);
const web3 = new Web3(provider);
```

The above code is a Web3 instance using a Polyjuice Web3 Provider. We will just call it "Polyjuice Web3" for short. We need to take this code and replace the existing Ethereum Web3 instance. In `app.tsx`, locate the existing Web3 instance, which should match the line below.

```js
const web3 = new Web3((window as any).ethereum);
```

Delete this line, and replace it with the Polyjuice Web3 code from above. Now our application is setup to communicate with Polyjuice using Web3!

### 5. Set High Gas Limit

Godwoken requires the gas limit to be set when sending transactions. This may not always be the case in the future, but it is a requirement for the current version on the Testnet.

To accomodate for this, we can make a simple change to default the gas limit to `6000000` for the user when they make transactions. In our project, this is all handled in the file `SimpleStorageWrapper.ts`. Open the file `~/projects/blockchain-workshop-ethereum-simple/src/lib/contracts/SimpleStorageWrapper.ts` in your editor.

First, we define a simple object that contains the `gas` property used by MetaMask.

```js
const DEFAULT_SEND_OPTIONS = {
    gas: 6000000
};
```

This can be added in the top region of the file, and we will be using this constant in several other places.

We will be adding it into the object passed to `send()` as the default values. For example, this code:

```js
this.contract.methods.set(value).send({
    from: fromAddress
});
```

is changed to:

```js
this.contract.methods.set(value).send({
    ...DEFAULT_SEND_OPTIONS,
    from: fromAddress
});
```

### 6. Display Polyjuice Address in Your Application

Every Ethereum address can be translated into a Polyjuice address on Nervos' Layer 2. This can be done using the `AddressTranslator` class. We will show the basic code here, but we will not cover the necessary changes for React to display it.

```js
import { AddressTranslator } from 'nervos-godwoken-integration';
```

We can then use the following code to find the Polyjuice address.

```js
const addressTranslator = new AddressTranslator();
const polyjuiceAddress = addressTranslator.ethAddressToGodwokenShortAddress(ethereumAddress);
```

### 7. View the Completed Godwoken Demo Application

We covered the main code changes required to port an Ethereum application to Nervos, but several small other changes were also necessary to display the new information in React. We didn't cover this since it is out of scope of this guide.

To view the final ported application you can checkout the [godwoken-simple](https://github.com/Kuzirashi/blockchain-workshop/tree/godwoken-simple) branch of the project using the commands below.

> Note: If you still have the Ethereum version of the project running, make sure to shut it down before launching the Godwoken version.

```sh
git clone https://github.com/Kuzirashi/blockchain-workshop.git -b godwoken-simple blockchain-workshop-godwoken-simple
cd blockchain-workshop-godwoken-simple
yarn
yarn build
yarn ui
```

This will start the UI server, and you can now open http://localhost:3000 in a browser. Once your browser is open, change your MetaMask network to `Godwoken Testnet`, which we setup earlier in this guide.

You can now try out the application running on the Godwoken Testnet!

<img src="https://raw.githubusercontent.com/Kuzirashi/gw-gitcoin-instruction/master/src/images/ported-confirm-deploy-sign.png" height="400" />

### Potential Errors and Solutions

- If you get a CORS error in your web browser's console, try searching your code for a Godwoken RPC URL that is **not** using `https`. Change any instances of `http://godwoken-testnet-web3-rpc.ckbapp.dev` to `https://godwoken-testnet-web3-rpc.ckbapp.dev`.

## Task Submission

To complete the tasks, add the following materials to a document on your Github and submit for review by the judges (include the link in your Gitcoin submission):

1. Screenshots or video of your application running on Godwoken.
2. Link to the GitHub repository with your application which has been ported to Godwoken. **This must be a different application than the one covered in this guide.**
3. If you deployed any smart contracts as part of this tutorial, please provide the transaction hash of the deployment transaction, the deployed contract address, and the ABI of the deployed smart contract. (Provide all in text format.)

### Web3.js

Instructions how to migrate your dApp if you're using web3.js can be found [here.](https://github.com/nervosnetwork/polyjuice-provider/blob/main/packages/ethers/README.md)

### Ethers.js

Instructions how to migrate your dApp if you're using ethers.js can be found [here.](https://github.com/nervosnetwork/polyjuice-provider/blob/main/packages/web3/README.md)

### Caveats

There are a number of small differences that can potentially impact your application and cause problems if you're not aware of them. A list of these differences can be found [here](https://github.com/nervosnetwork/godwoken-polyjuice/blob/main/docs/EVM-compatible.md).

If you need to use `ecrecover` you will have to modify your smart-contracts to do an internal system call using assembly. You can find documentation for that syscall [here](https://github.com/nervosnetwork/godwoken-polyjuice/blob/main/docs/Addition-Features.md), there is an [example code](https://github.com/RetricSu/godwoken-polyjuice-compatibility-examples#compatibility).

### More Examples

If you need more examples of working Nervos' Layer 2 applications you can take a look at the following repositories:

1. [Head or Tail bet application](https://github.com/Kuzirashi/blockchain-workshop/tree/godwoken)
2. [godwoken-polyjuice-compatibility-examples repository](https://github.com/RetricSu/godwoken-polyjuice-compatibility-examples)

## Helpful Links

- [Introduction](https://github.com/Kuzirashi/gw-gitcoin-instruction/blob/master/src/introduction/introduction.md)
- [Task Setup and Requirements](https://github.com/Kuzirashi/gw-gitcoin-instruction/tree/master/src/task-setup-and-requirements)
- [Discord](https://discord.com/invite/AqGTUE9)
- [Broaden the Spectrum](https://gitcoin.co/hackathon/nervos/onboard)
- [Bounties List](https://gitcoin.co/hackathon/nervos/)
- [Content Directory](https://github.com/Kuzirashi/gw-gitcoin-instruction)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 2 weeks, 5 days from now.
Please review their action plans below:


**1) [timfaner](https://gitcoin.co/timfaner) has started work.**

https://github.com/timfaner/gitcoin-task/tree/main/task7
Task7 submit
**2) [antonyip](https://gitcoin.co/antonyip) has started work.**

task7
**3) [x777](https://gitcoin.co/x777) has started work.**

Port An Existing Ethereum DApp To Polyjuice
**4) [soptq](https://gitcoin.co/soptq) has started work.**

I will port a todo list dapp to the polyjuice.
**5) [l-kh](https://gitcoin.co/l-kh) has started work.**

This is my submission for Nervos Hackathon Gitcoin7
**6) [phungnm](https://gitcoin.co/phungnm) has started work.**

Port An Existing Ethereum DApp To Polyjuice
**7) [vinhbhn](https://gitcoin.co/vinhbhn) has started work.**

Port An Existing Ethereum DApp To Polyjuice
**8) [ubinix-warun](https://gitcoin.co/ubinix-warun) has started work.**

1. Screenshots or video of your application running on Godwoken.
2. Link to the GitHub repository with your application which has been ported to Godwoken. This must be a different application than the one covered in this guide.
3. If you deployed any smart contracts as part of this tutorial, please provide the transaction hash of the deployment transaction, the deployed contract address, and the ABI of the deployed smart contract. (Provide all in text format.)
**9) [sidduhere](https://gitcoin.co/sidduhere) has started work.**

Task submission
**10) [orangemio](https://gitcoin.co/orangemio) has started work.**

https://github.com/orangemio/blockchain-workshop
**11) [ayush20](https://gitcoin.co/ayush20) has started work.**

Port a Ethereum dapp to Polyjuice
**12) [assafom](https://gitcoin.co/assafom) has started work.**

Step 7.
**13) [maxx6262](https://gitcoin.co/maxx6262) has started work.**

Gitcoin: 7) Port An Existing Ethereum DApp To Polyjuice
**14) [nomicid](https://gitcoin.co/nomicid) has started work.**

Working on Task7
**15) [ndduc-fpt](https://gitcoin.co/ndduc-fpt) has started work.**

Gitcoin: 7) Port An Existing Ethereum DApp To Polyjuice
**16) [da3dalus88](https://gitcoin.co/da3dalus88) has started work.**

Nervos Network hackathon
**17) [huutrung](https://gitcoin.co/huutrung) has started work.**

task7
**18) [lehoi2195](https://gitcoin.co/lehoi2195) has started work.**

Port An Existing Ethereum DApp To Polyjuice
**19) [dex68](https://gitcoin.co/dex68) has started work.**

ok
**20) [raiden1411](https://gitcoin.co/raiden1411) has started work.**

Nervos Port An Existing Ethereum DApp To Polyjuice
**21) [sking789](https://gitcoin.co/sking789) has started work.**

repo for ckb gitcoin hackathons
**22) [andermarce](https://gitcoin.co/andermarce) has started work.**

Task7
**23) [anhnt4288](https://gitcoin.co/anhnt4288) has started work.**

Nervous hackathon gitcoin
**24) [fabioktldaf](https://gitcoin.co/fabioktldaf) has started work.**

Port An Existing Ethereum DApp To Polyjuice
**25) [stwith](https://gitcoin.co/stwith) has started work.**

nervos
**26) [tharunrai14](https://gitcoin.co/tharunrai14) has started work.**

Nervos hacakthon task 07
**27) [duylinh196tb](https://gitcoin.co/duylinh196tb) has started work.**

Nervos gitcoin task 7
**28) [shinyhunters](https://gitcoin.co/shinyhunters) has started work.**

Nervos task7
**29) [mistakeone](https://gitcoin.co/mistakeone) has started work.**

task 7
**30) [ttnguyendev](https://gitcoin.co/ttnguyendev) has started work.**

A project for hackathon
**31) [preet-aulokh](https://gitcoin.co/preet-aulokh) has started work.**

Migrating Ethereum dapp to Polyjuice
**32) [celoaken](https://gitcoin.co/celoaken) has started work.**

Write and deploy a new DApp
**33) [baoanh1310](https://gitcoin.co/baoanh1310) has started work.**

Port an existing Ethereum dApp to Polyjuice
**34) [xinbadev](https://gitcoin.co/xinbadev) has started work.**

gitcoin7
**35) [nishuzumi](https://gitcoin.co/nishuzumi) has started work.**

nervos
**36) [quocson95](https://gitcoin.co/quocson95) has started work.**

nervos-task 7
**37) [pulth](https://gitcoin.co/pulth) has started work.**

NervosHackathon Task7 Submission
**38) [chiragbadhe](https://gitcoin.co/chiragbadhe) has started work.**

Gitcoin: 7) Port An Existing Ethereum DApp To Polyjuice
**39) [rzbck](https://gitcoin.co/rzbck) has started work.**

Gitcoin: 7) Port An Existing Ethereum DApp To Polyjuice
**40) [sondq](https://gitcoin.co/sondq) has started work.**

nervos 7
**41) [niedhui](https://gitcoin.co/niedhui) has started work.**

Task 08
**42) [quocson100](https://gitcoin.co/quocson100) has started work.**

Task 7
**43) [carmen0208](https://gitcoin.co/carmen0208) has started work.**

Port An Existing Ethereum DApp To Polyjuice
**44) [tn024687](https://gitcoin.co/tn024687) has started work.**

Port To Do DApp To Polyjuice
**45) [nicky-ru](https://gitcoin.co/nicky-ru) has started work.**

In my project I will try to deploy a well-known Rubixi contract on Nervos Layer 2.
**46) [domix14](https://gitcoin.co/domix14) has started work.**

nervos task 7
**47) [selimkafa](https://gitcoin.co/selimkafa) has started work.**

nervos7
**48) [bayou020](https://gitcoin.co/bayou020) has started work.**

task 7
**49) [hodlrtodlrfarmr](https://gitcoin.co/hodlrtodlrfarmr) has started work.**

Gitcoin: 7 Submission
**50) [sgnovo8](https://gitcoin.co/sgnovo8) has started work.**

Port An Existing Ethereum DApp To Polyjuice
**51) [pfed-prog](https://gitcoin.co/pfed-prog) has started work.**

working on project 7 nervos
**52) [dikky88](https://gitcoin.co/dikky88) has started work.**

ckb dikky task7
**53) [iktumi](https://gitcoin.co/iktumi) has started work.**

Port An Existing Ethereum DApp To Polyjuice
**54) [cryptohong](https://gitcoin.co/cryptohong) has started work.**

Nervos Hackathon
**55) [gabrycina](https://gitcoin.co/gabrycina) has started work.**

Nervos Hackaton Task 7
**56) [ben-razor](https://gitcoin.co/ben-razor) has started work.**

Port An Existing Ethereum DApp To Polyjuice
**57) [gdawson771](https://gitcoin.co/gdawson771) has started work.**

Im a 3rd year CS&AI student at edinburgh excited to begin blockchain development
**58) [cocdap](https://gitcoin.co/cocdap) has started work.**

This is task 7
**59) [jpeterd](https://gitcoin.co/jpeterd) has started work.**

Porting exisiting Ethereum Dapp.
**60) [knnlrts](https://gitcoin.co/knnlrts) has started work.**

Nervos hackaton: task 7
**61) [hg00](https://gitcoin.co/hg00) has started work.**

Porting ETH contract over
**62) [jackieyewang](https://gitcoin.co/jackieyewang) has started work.**

https://github.com/Jackieyewang/Nervos-gitcoin/tree/main/task7
**63) [rkr-dev](https://gitcoin.co/rkr-dev) has started work.**

Port an Existing Ethereum dApp to Polyjuice
**64) [long-blade](https://gitcoin.co/long-blade) has started work.**

Gitcoin: 7) Port an Existing Ethereum dApp to Polyjuice
**65) [andithemudkip](https://gitcoin.co/andithemudkip) has started work.**

Task 8
**66) [superheavytoken](https://gitcoin.co/superheavytoken) has started work.**

Stage 8 Hackathon Nervos
**67) [scentlxss](https://gitcoin.co/scentlxss) has started work.**

nervos task 7
**68) [raghav1997](https://gitcoin.co/raghav1997) has started work.**

task 7
**69) [berkayermis](https://gitcoin.co/berkayermis) has started work.**

Submissions of Task 7
**70) [theyashmhatre](https://gitcoin.co/theyashmhatre) has started work.**

Task 7
**71) [happylolonly](https://gitcoin.co/happylolonly) has started work.**

https://github.com/happylolonly/nervos-hackathon/tree/main/task-7
**72) [terryboy](https://gitcoin.co/terryboy) has started work.**

task7
**73) [tanishqdsharma](https://gitcoin.co/tanishqdsharma) has started work.**

Deploy Dapp
**74) [cryptostwich](https://gitcoin.co/cryptostwich) has started work.**

Nervos Hackathon
**75) [jcervante](https://gitcoin.co/jcervante) has started work.**

Port An Existing Ethereum DApp To Polyjuice
**76) [hdbhandari](https://gitcoin.co/hdbhandari) has started work.**

Gitcoin: 7) Port An Existing Ethereum DApp To Polyjuice
**77) [ircrp](https://gitcoin.co/ircrp) has started work.**

Create a simple token application which can be deployed and interacted on the Nervos Layer 2 network
**78) [sabaiprimo](https://gitcoin.co/sabaiprimo) has started work.**

Project for Nervos Hackathon task7 (Airdrop application)
**79) [andrewpsp](https://gitcoin.co/andrewpsp) has started work.**

Welcome to the yard
**80) [ririen](https://gitcoin.co/ririen) has started work.**

This project is mostly fun. For the purposes of issue, I created a DApp in the Nuxt.js framework. Unique fancy names can be generated in the application.
**81) [kidneyweakx](https://gitcoin.co/kidneyweakx) has started work.**

Port a Dapp to Polyjuice anf record a Demo Video.
**82) [xcshuan](https://gitcoin.co/xcshuan) has started work.**

Gitcoin: 7) Port An Existing Ethereum DApp To Polyjuice
**83) [gentlebutapril](https://gitcoin.co/gentlebutapril) has started work.**

task7
**84) [hhuuaa](https://gitcoin.co/hhuuaa) has started work.**

do
**85) [nhaga](https://gitcoin.co/nhaga) has started work.**

nervos task 7
**86) [huazhouwang](https://gitcoin.co/huazhouwang) has started work.**

a
**87) [liushooter](https://gitcoin.co/liushooter) has started work.**

ckb
**88) [criusxx](https://gitcoin.co/criusxx) has started work.**

nervos-gitcoin7
**89) [arnaudatcomet](https://gitcoin.co/arnaudatcomet) has started work.**

HKT Nervos #7
**90) [jinusean](https://gitcoin.co/jinusean) has started work.**

Gitcoin: 7) Port An Existing Ethereum DApp To Polyjuice
**91) [chaitanyasjoshi](https://gitcoin.co/chaitanyasjoshi) has started work.**

Ported An Existing Ethereum DApp To Polyjuice
**92) [jiddukrish](https://gitcoin.co/jiddukrish) has started work.**

Gitcoin 7
**93) [playflycode](https://gitcoin.co/playflycode) has started work.**

fine
**94) [kizna1ver](https://gitcoin.co/kizna1ver) has started work.**

task7
**95) [ody123](https://gitcoin.co/ody123) has started work.**

task7
**96) [lqne](https://gitcoin.co/lqne) has started work.**

polyjuice-dapp-port
**97) [techmeupnl](https://gitcoin.co/techmeupnl) has started work.**

Hackathon
**98) [el-tumero](https://gitcoin.co/el-tumero) has started work.**

Eight encounter with Nervos :)
**99) [stgllr](https://gitcoin.co/stgllr) has started work.**

nervos-gitcoin
**100) [cito-lito](https://gitcoin.co/cito-lito) has started work.**

nervos
**101) [fsy412](https://gitcoin.co/fsy412) has started work.**

task 7
**102) [cuongtuanvu](https://gitcoin.co/cuongtuanvu) has started work.**

Port An Existing Ethereum DApp To Polyjuice
**103) [alexfedotovqq](https://gitcoin.co/alexfedotovqq) has started work.**

let's go
**104) [sunydev](https://gitcoin.co/sunydev) has started work.**

Port an Existing Ethereum dApp to Polyjuice
**105) [amrosaeed](https://gitcoin.co/amrosaeed) has started work.**

Nervos Hackathon
**106) [cultless](https://gitcoin.co/cultless) has started work.**

07_Nervos - Port an Existing Ethereum dApp to Polyjuice
**107) [crispysandwhich](https://gitcoin.co/crispysandwhich) has started work.**

Basically using a project from crypto zombie and putting it on the godwoken testnet
**108) [schrubitteflau](https://gitcoin.co/schrubitteflau) has started work.**

Hackathon: Nervos - Broaden the Spectrum
Gitcoin: 7) Port an Existing Ethereum dApp to Polyjuice
**109) [mirrormirage0](https://gitcoin.co/mirrormirage0) has started work.**

Deploy a custom ERC20 Token Deployer DApp on Godwoken
**110) [helmihidzir](https://gitcoin.co/helmihidzir) has started work.**

Port An Existing Ethereum DApp To Polyjuice
**111) [larricane](https://gitcoin.co/larricane) has started work.**

Gitcoin:  7)
**112) [oceansxq](https://gitcoin.co/oceansxq) has started work.**

nervos-gitcoin-7
**113) [7kms](https://gitcoin.co/7kms) has started work.**

Port an Existing Ethereum dApp to Polyjuice
**114) [elfelf2](https://gitcoin.co/elfelf2) has started work.**

gitcoin7
**115) [puppetmaster66](https://gitcoin.co/puppetmaster66) has started work.**

Task 7

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/nervosnetwork/grants/8/100026214).

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Work for 7500.0 CKB (75.75 USD @ $0.01/CKB) has been submitted by__:

 1. [@antonyip](https://gitcoin.co/antonyip)
 2. [@soptq](https://gitcoin.co/soptq)
 3. [@l-kh](https://gitcoin.co/l-kh)
 4. [@timfaner](https://gitcoin.co/timfaner)
 5. [@vinhbhn](https://gitcoin.co/vinhbhn)
 6. [@sidduhere](https://gitcoin.co/sidduhere)
 7. [@orangemio](https://gitcoin.co/orangemio)
 8. [@raiden1411](https://gitcoin.co/raiden1411)
 9. [@lehoi2195](https://gitcoin.co/lehoi2195)
 10. [@sking789](https://gitcoin.co/sking789)
 11. [@stwith](https://gitcoin.co/stwith)
 12. [@ubinix-warun](https://gitcoin.co/ubinix-warun)
 13. [@andermarce](https://gitcoin.co/andermarce)
 14. [@duylinh196tb](https://gitcoin.co/duylinh196tb)
 15. [@shinyhunters](https://gitcoin.co/shinyhunters)
 16. [@ttnguyendev](https://gitcoin.co/ttnguyendev)
 17. [@ronalddas](https://gitcoin.co/ronalddas)
 18. [@orhanors](https://gitcoin.co/orhanors)
 19. [@phungnm](https://gitcoin.co/phungnm)
 20. [@celoaken](https://gitcoin.co/celoaken)
 21. [@tech-engine](https://gitcoin.co/tech-engine)
 22. [@xinbadev](https://gitcoin.co/xinbadev)
 23. [@nishuzumi](https://gitcoin.co/nishuzumi)
 24. [@mistakeone](https://gitcoin.co/mistakeone)
 25. [@quocson95](https://gitcoin.co/quocson95)
 26. [@sondq](https://gitcoin.co/sondq)
 27. [@niedhui](https://gitcoin.co/niedhui)
 28. [@rzbck](https://gitcoin.co/rzbck)
 29. [@quocson100](https://gitcoin.co/quocson100)
 30. [@carmen0208](https://gitcoin.co/carmen0208)
 31. [@tn024687](https://gitcoin.co/tn024687)
 32. [@nicky-ru](https://gitcoin.co/nicky-ru)
 33. [@domix14](https://gitcoin.co/domix14)
 34. [@selimkafa](https://gitcoin.co/selimkafa)
 35. [@hodlrtodlrfarmr](https://gitcoin.co/hodlrtodlrfarmr)
 36. [@samet1998](https://gitcoin.co/samet1998)
 37. [@dex68](https://gitcoin.co/dex68)
 38. [@da3dalus88](https://gitcoin.co/da3dalus88)
 39. [@pfed-prog](https://gitcoin.co/pfed-prog)
 40. [@cryptohong](https://gitcoin.co/cryptohong)
 41. [@carljungvibe](https://gitcoin.co/carljungvibe)
 42. [@gabrycina](https://gitcoin.co/gabrycina)
 43. [@ben-razor](https://gitcoin.co/ben-razor)
 44. [@postman56](https://gitcoin.co/postman56)
 45. [@cocdap](https://gitcoin.co/cocdap)
 46. [@knnlrts](https://gitcoin.co/knnlrts)
 47. [@jackieyewang](https://gitcoin.co/jackieyewang)
 48. [@bayou020](https://gitcoin.co/bayou020)
 49. [@nervos2021](https://gitcoin.co/nervos2021)
 50. [@andithemudkip](https://gitcoin.co/andithemudkip)
 51. [@jpeterd](https://gitcoin.co/jpeterd)
 52. [@superheavytoken](https://gitcoin.co/superheavytoken)
 53. [@pulth](https://gitcoin.co/pulth)
 54. [@assafom](https://gitcoin.co/assafom)
 55. [@scentlxss](https://gitcoin.co/scentlxss)
 56. [@theyashmhatre](https://gitcoin.co/theyashmhatre)
 57. [@berkayermis](https://gitcoin.co/berkayermis)
 58. [@happylolonly](https://gitcoin.co/happylolonly)
 59. [@terryboy](https://gitcoin.co/terryboy)
 60. [@tanishqdsharma](https://gitcoin.co/tanishqdsharma)
 61. [@cryptostwich](https://gitcoin.co/cryptostwich)
 62. [@jcervante](https://gitcoin.co/jcervante)
 63. [@anhnt4288](https://gitcoin.co/anhnt4288)
 64. [@ircrp](https://gitcoin.co/ircrp)
 65. [@sabaiprimo](https://gitcoin.co/sabaiprimo)
 66. [@andrewpsp](https://gitcoin.co/andrewpsp)
 67. [@ririen](https://gitcoin.co/ririen)
 68. [@homerosland](https://gitcoin.co/homerosland)
 69. [@kidneyweakx](https://gitcoin.co/kidneyweakx)
 70. [@gentlebutapril](https://gitcoin.co/gentlebutapril)
 71. [@hhuuaa](https://gitcoin.co/hhuuaa)
 72. [@nhaga](https://gitcoin.co/nhaga)
 73. [@xcshuan](https://gitcoin.co/xcshuan)
 74. [@huazhouwang](https://gitcoin.co/huazhouwang)
 75. [@liushooter](https://gitcoin.co/liushooter)
 76. [@criusxx](https://gitcoin.co/criusxx)
 77. [@arnaudatcomet](https://gitcoin.co/arnaudatcomet)
 78. [@jinusean](https://gitcoin.co/jinusean)
 79. [@chaitanyasjoshi](https://gitcoin.co/chaitanyasjoshi)
 80. [@bulentbozdag](https://gitcoin.co/bulentbozdag)
 81. [@x777](https://gitcoin.co/x777)
 82. [@jiddukrish](https://gitcoin.co/jiddukrish)
 83. [@lqne](https://gitcoin.co/lqne)
 84. [@fabioktldaf](https://gitcoin.co/fabioktldaf)
 85. [@el-tumero](https://gitcoin.co/el-tumero)
 86. [@alanfreud](https://gitcoin.co/alanfreud)
 87. [@techmeupnl](https://gitcoin.co/techmeupnl)
 88. [@stgllr](https://gitcoin.co/stgllr)
 89. [@cito-lito](https://gitcoin.co/cito-lito)
 90. [@icaroo](https://gitcoin.co/icaroo)
 91. [@fsy412](https://gitcoin.co/fsy412)
 92. [@alexfedotovqq](https://gitcoin.co/alexfedotovqq)
 93. [@sunydev](https://gitcoin.co/sunydev)
 94. [@amrosaeed](https://gitcoin.co/amrosaeed)
 95. [@crispysandwhich](https://gitcoin.co/crispysandwhich)
 96. [@kizna1ver](https://gitcoin.co/kizna1ver)
 97. [@taylorkhalifa](https://gitcoin.co/taylorkhalifa)
 98. [@schrubitteflau](https://gitcoin.co/schrubitteflau)
 99. [@ody123](https://gitcoin.co/ody123)
 100. [@mirrormirage0](https://gitcoin.co/mirrormirage0)
 101. [@wizardlog](https://gitcoin.co/wizardlog)
 102. [@oceansxq](https://gitcoin.co/oceansxq)
 103. [@elfelf2](https://gitcoin.co/elfelf2)



@mattquinn-nervos please take a look at the submitted work:
* [PR](https://github.com/elfelf2/gitcoin-nervos/tree/main/gitcoin7) by @elfelf2
* [PR](https://github.com/oceanSxq/nervos-gitcoin/tree/main/ckb-task-7) by @oceansxq
* [PR](https://github.com/wizardlog/gitcoin-nervos-hackacton/tree/master/gitcoin-7) by @wizardlog
* [PR](https://github.com/mirrormirage0/nervos/tree/main/Hackathon-Task-7-TokenDeployerPorted) by @mirrormirage0
* [PR](https://github.com/ody123/nervos-grants/tree/master/Gitcoin%207)%20Port%20An%20Existing%20Ethereum%20DApp%20To%20Polyjuice) by @ody123
* [PR](https://github.com/Schrubitteflau/Nervos-Hackaton/tree/master/7-Ethereum-DApp-Polyjuice) by @schrubitteflau
* [PR](https://github.com/taylorkhalifa/nervos-gitcoin-hackathon/tree/master/gitcoin-7)%20Porting%20Dapp) by @taylorkhalifa
* [PR](https://github.com/Kizna1ver/CKBProjects/tree/main/task7) by @kizna1ver
* [PR](https://github.com/shpintz/NervosHackathon) by @crispysandwhich
* [PR](https://github.com/amrosaeed/Nervos-Hackathon/tree/master/Task%207) by @amrosaeed
* [PR](https://github.com/sunydev/Nervos-Network-Hackathon/blob/main/task7/Gitcoin:%207)%20Port%20an%20Existing%20Ethereum%20dApp%20to%20Polyjuice.md) by @sunydev
* [PR](https://github.com/AlexFedotovqq/nervos/tree/main/task7) by @alexfedotovqq
* [PR](https://github.com/fsy412/Nervos-Gitcoin-Hackathon/tree/main/task_7) by @fsy412
* [PR](https://github.com/icaroo/nervosTask7) by @icaroo
* [PR](https://github.com/cito-lito/nervos_hackathon/tree/main/task_07) by @cito-lito
* [PR](https://github.com/stgllr/nervos-gitcoin/tree/main/Assets) by @stgllr
* [PR](https://github.com/stevenhva/NervosNetworkHackathon/blob/main/Task_7.md) by @techmeupnl
* [PR](https://github.com/alanfreud/Nervos-gitcoin-projects/tree/master/gitcoin-7) by @alanfreud
* [PR](https://github.com/el-tumero/nervos7) by @el-tumero
* [PR](https://github.com/fabioktldaf/nervosnetwork/tree/main/07) by @fabioktldaf
* [PR](https://github.com/lqne/polyjuice-dapp-port) by @lqne
* [PR](https://github.com/jiddukrish/Nervos-gitcoin/tree/master/task7) by @jiddukrish
* [PR](https://github.com/x777/Nervos-Gitconin-Hackaton) by @x777
* [PR](https://github.com/bulentbozdag/nervos-testnet/tree/main/task-7) by @bulentbozdag
* [PR](https://github.com/chaitanyasjoshi/nervos-hackathon/tree/main/Task%207) by @chaitanyasjoshi
* [PR](https://github.com/jinusean/nervosness/tree/main/task-7) by @jinusean
* [PR](https://github.com/arnaudatcomet/nervos/tree/main/task-7) by @arnaudatcomet
* [PR](https://github.com/criusxx/nervos-gitcoin/tree/main/task7) by @criusxx
* [PR](https://github.com/liushooter/ckb-bounty/tree/main/task08) by @liushooter
* [PR](https://github.com/huazhouwang/gitcoin_ckb/tree/main/task7) by @huazhouwang
* [PR](https://github.com/xcshuan/nervos-gitcoin/tree/main/gitcoin-7) by @xcshuan
* [PR](https://github.com/nhaga/nervos-hack/tree/main/task7) by @nhaga
* [PR](https://github.com/hhuuaa/gitcoin_ckb/tree/main/task7) by @hhuuaa
* [PR](https://github.com/GentleButApril/nervos-gitcoin/tree/main/task-7) by @gentlebutapril
* [PR](https://github.com/kidneyweakx/nervos-gitcoin/tree/main/task-7) by @kidneyweakx
* [PR](https://github.com/homerosland/nervos-gitcoin-hack/tree/master/task7) by @homerosland
* [PR](https://github.com/ririen/broaden-the-spectrum/tree/main/Ethereum%20DApp) by @ririen
* [PR](https://github.com/andrewpsp/gityard/tree/master/Gitcoin-7-Port-An-Existing-Ethereum-DApp-To-Polyjuice) by @andrewpsp
* [PR](https://github.com/sabaiprimo/Nervos-Hackathon/tree/main/task7) by @sabaiprimo
* [PR](https://github.com/ircrp/gitcoin-nervosnetwork/blob/main/7/README.md) by @ircrp
* [PR](https://github.com/anhnt4288/nervous-hackathon/tree/master/task-7) by @anhnt4288
* [PR](https://github.com/jcervante/CBKTASKs/tree/main/TASK07) by @jcervante
* [PR](https://github.com/cryptostwich/nervos-work/blob/main/task7) by @cryptostwich
* [PR](https://github.com/TanishqDsharma/nervous-network-task-7) by @tanishqdsharma
* [PR](https://github.com/terryboy/nervos-gitcoin/tree/main/TASK-7) by @terryboy
* [PR](https://github.com/happylolonly/nervos-hackathon/tree/main/task-7) by @happylolonly
* [PR](https://github.com/berkayermis/Nervos-Gitcoin-Hackathon/tree/main/t7) by @berkayermis
* [PR](https://github.com/theyashmhatre/nervos-hackathon/tree/main/7) by @theyashmhatre
* [PR](https://github.com/scentlxss/gitcoin_nervos_tasks/tree/main/task7) by @scentlxss
* [PR](https://github.com/assafom/hackathon-nervosnetwork/tree/main/step-7) by @assafom
* [PR](https://github.com/pulth/nervos-hackathon/tree/main/task7) by @pulth
* [PR](https://github.com/superheavytoken/Gitcoin-7-Port-An-Existing-Ethereum-DApp-To-Polyjuice) by @superheavytoken
* [PR](https://github.com/JPeterD/Gitcoin-Nervos/blob/main/Task%207/README.md) by @jpeterd
* [PR](https://github.com/andithemudkip/nervos/tree/main/task7) by @andithemudkip
* [PR](https://github.com/nervos2021/gitcoin_nervos/tree/main/task7) by @nervos2021
* [PR](https://github.com/bayou020/nervosnetwork/tree/main/task7) by @bayou020
* [PR](https://github.com/Jackieyewang/Nervos-gitcoin/tree/main/task7) by @jackieyewang
* [PR](https://github.com/knnlrts/nervos-hackaton/tree/main/task-7) by @knnlrts
* [PR](https://github.com/CocDap/nervos-hackathon/tree/master/task_7) by @cocdap
* [PR](https://github.com/PostMan56/Nervos-BTS-Gitcoin/tree/main/Nervos-07) by @postman56
* [PR](https://github.com/ben-razor/crypto-funk-app) by @ben-razor
* [PR](https://github.com/gabrycina/nervos/tree/main/task7) by @gabrycina
* [PR](https://github.com/carljungvibe/nervos-gitcoin/tree/master/task7) by @carljungvibe
* [PR](https://github.com/cryptoHong/Nervos-Hackathon/tree/main/task7) by @cryptohong
* [PR](https://github.com/Pfed-prog/gitcoin_nervos/tree/main/task7) by @pfed-prog
* [PR](https://github.com/da3dalus88/nervosHackathon/blob/main/tasks/task7/task7.md) by @da3dalus88
* [PR](https://github.com/Dex68/nervos_gitcoin/blob/main/task7/README.md) by @dex68
* [PR](https://github.com/samet1998/7-Nervos) by @samet1998
* [PR](https://github.com/hodlrtodlrfarmr/gitcoin_7) by @hodlrtodlrfarmr
* [PR](https://github.com/selimkafa/nervos-gitcoin/tree/main/task7) by @selimkafa
* [PR](https://github.com/Domix14/Hackathon_Nervos/tree/main/7) by @domix14
* [PR](https://github.com/nicky-ru/nervos/tree/main/gitcoin7) by @nicky-ru
* [PR](https://github.com/tn024687/nervos-hackathon/tree/master/task-7) by @tn024687
* [PR](https://github.com/carmen0208/gitcoin-nervos/tree/main/07) by @carmen0208
* [PR](https://github.com/quocson100/task/tree/master/task7) by @quocson100
* [PR](https://github.com/Rzbck/Rzbck-Nervos/tree/main/Gitcoin:%207-%20Port%20An%20Existing%20Ethereum%20DApp%20To%20Polyjuice) by @rzbck
* [PR](https://github.com/niedhui/nervos-hackaton/tree/main/07) by @niedhui
* [PR](https://github.com/sondq/nervos/tree/master/task7) by @sondq
* [PR](https://github.com/quocson95/nervos-task/tree/master/task7) by @quocson95
* [PR](https://github.com/mistakeone/nervos_first_try/blob/master/7%20task/README.MD) by @mistakeone
* [PR](https://github.com/nishuzumi/nervos-hackthon/tree/master/task7) by @nishuzumi
* [PR](https://github.com/xinbaDev/nervos-gitcoin#gitcoin-7-port-an-existing-ethereum-dapp-to-polyjuice) by @xinbadev
* [PR](https://github.com/Tech-engine/nervos-network/tree/main/task7) by @tech-engine
* [PR](https://github.com/celoaken/gitcoin-nervos/tree/main/task7) by @celoaken
* [PR](https://github.com/phungnm/nervos-gitcoin-hackathon/tree/master/task-7) by @phungnm
* [PR](https://github.com/orhanors/nervos-hackhaton/tree/master/task7) by @orhanors
* [PR](https://github.com/ronalddas/gitcoin-nervos/tree/master/task7) by @ronalddas
* [PR](https://github.com/TTNguyenDev/Hackathon-Nervos/tree/main/task_7) by @ttnguyendev
* [PR](https://github.com/shinyhunters/gitcoin_project0/tree/main/task7) by @shinyhunters
* [PR](https://github.com/duylinh196tb/Nervos_7) by @duylinh196tb
* [PR](https://github.com/andermarce/nervos-hackaton/tree/main/task7) by @andermarce
* [PR](https://github.com/ubinix-warun/gitcoin-nervos-broaden-the-spectrum/blob/master/my-submission/task-7/README.md) by @ubinix-warun
* [PR](https://github.com/stwith/NervosHackathon/tree/main/task7) by @stwith
* [PR](https://github.com/sking789/ckb_gitcoin_hackathons/tree/main/task7) by @sking789
* [PR](https://github.com/lehoi2195/blockchain-workshop) by @lehoi2195
* [PR](https://github.com/Raiden1411/NervosHackathon_7/blob/main/README.md) by @raiden1411
* [PR](https://github.com/orangemio/blockchain-workshop) by @orangemio
* [PR](https://github.com/sidduHERE/Nervos/tree/main/Task7) by @sidduhere
* [PR](https://github.com/vinhbhn/nervos-gitcoin/tree/main/task-7) by @vinhbhn
* [PR](https://github.com/timfaner/gitcoin-task/tree/main/task7) by @timfaner
* [PR](https://github.com/L-KH/NervousHackathon/tree/main/Gitcoin7) by @l-kh
* [PR](https://github.com/Soptq/2021-Nervos-Broaden-the-Spectrum-task7) by @soptq
* [PR](https://github.com/antonyip/nervos-gitcoin/tree/main/task-7) by @antonyip

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/nervosnetwork/grants/8/100026214)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/nervosnetwork/grants/8/100026214).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://discord.gg/gitcoin/'>Gitcoin's Discord</a>
* $2,229,860.60 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by shinyhunters** [comment]
HELLO, I HAVE SUBMITTED THE TASK7 BUT STILL I HAVE NOT RECIEVED THE REWARDS, PLEASE HELP
